services/header/rocketjoe/services/python_sandbox/celery.cpp

```diff
@@ -1,50 +1,215 @@
+        //TODO: fix memory task_handler leak
+    }
+
+    auto task::apply(py::args args, py::kwargs kwargs) -> eager_result const {
+        auto task_id = boost::uuids::random_generator()();
+
+        return eager_result{std::move(task_id), task_handler(*args, **kwargs)};
+    }
+
+    auto task::operator()(py::args args, py::kwargs kwargs) -> eager_result const {
+        return apply(std::move(args), std::move(kwargs));
+    }
+
+    auto task::apply_async(py::args args, py::kwargs kwargs) -> async_result const {
+        auto task_id = boost::uuids::random_generator()();
+
+        app->send_task(task_name, std::move(args), std::move(kwargs), task_id);
+
+        return async_result{std::move(task_id), task_name, app};
+    }
+
+    task_pending::task_pending(std::string task_name, py::args args, py::kwargs kwargs, boost::uuids::uuid task_id, std::size_t priority)
+        : task_name{std::move(task_name)}
+        , args{std::move(args)}
+        , kwargs{std::move(kwargs)}
+        , task_id{std::move(task_id)}
+        , priority{priority} {}
+
+    bool task_pending::operator<(const task_pending &rhs) const {
+        return priority < rhs.priority;
+    }
+
+    task_answer::task_answer(py::object ret_val, std::size_t priority)
+        : ret_val{std::move(ret_val)}
+        , priority{priority} {}
+
+    bool task_answer::operator<(const task_answer &rhs) const {
+        return priority < rhs.priority;
+    }
+
+    worker::worker(boost::intrusive_ptr<celery> app) : app{std::move(app)} {}
+
+    auto worker::self() -> boost::intrusive_ptr<worker> const {
+        return boost::intrusive_ptr<worker>{this};
+    }
+
+    auto worker::poll() -> void {
+        auto task_pending = app->pop_top_task_pending();
+
+        if(!task_pending) {
+            throw py::stop_iteration();
+        }
+
+        auto task_handler = app->top_task_handler(std::move(task_pending->task_name));
+
+        if(!task_handler) {
+            return;
+        }
+
+        auto result = (*task_handler)(*task_pending->args, **task_pending->kwargs);
+
+        app->push_task_answer(std::move(task_pending->task_id), {std::move(result), 0});
+    }
+
+    auto celery::create_task(py::args args, py::kwargs kwargs) -> std::function<task(py::function)> const {
+        boost::intrusive_ptr<celery> self{this};
+
+        return [self{std::move(self)}](auto task_handler) {
+            auto module_name = py::cast<std::string>(*task_handler.attr("__module__"));
+            auto qualified_name = py::cast<std::string>(*task_handler.attr("__qualname__"));
+            auto task_name = module_name + "." + qualified_name;
+
+            return task{std::move(task_name), std::move(task_handler), std::move(self)};
+        };
+    }
+
+    auto celery::send_task(std::string name, py::args args, py::kwargs kwargs, boost::uuids::uuid task_id) -> void {
+        push_task_pending({std::move(name), std::move(args), std::move(kwargs), std::move(task_id), 0});
```

Comment:
It would be nice to have a constant and not a magic number.  constexpr const int priority_zero = 0 ;